File: packages/sdk-python/src/clawtoken/node.py
"""Node API — status, peers, config, sync."""

from __future__ import annotations

import time
from typing import Any

from clawtoken.http import AsyncHttpClient, HttpClient


class NodeApi:
    """Synchronous Node API."""

    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def get_status(self) -> dict[str, Any]:
        """Get node status (DID, sync state, block height, peers)."""
        return self._http.get("/api/node/status")

    def get_peers(self, *, limit: int | None = None, offset: int | None = None) -> dict[str, Any]:
        """List connected peers."""
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset
        return self._http.get("/api/node/peers", params or None)

    def get_config(self) -> dict[str, Any]:
        """Get node configuration."""
        return self._http.get("/api/node/config")

    def wait_for_sync(
        self,
        *,
        interval: float = 2.0,
        timeout: float = 60.0,
    ) -> dict[str, Any]:
        """Poll ``get_status`` until the node reports ``synced=True``.

        Raises ``TimeoutError`` if *timeout* seconds elapse.
        """
        deadline = time.monotonic() + timeout
        while True:
            status = self.get_status()
            if status.get("synced"):
                return status
            if time.monotonic() >= deadline:
                raise TimeoutError(f"Node did not sync within {timeout}s")
            time.sleep(interval)


class AsyncNodeApi:
    """Asynchronous Node API."""

    def __init__(self, http: AsyncHttpClient) -> None:
        self._http = http

    async def get_status(self) -> dict[str, Any]:
        return await self._http.get("/api/node/status")

    async def get_peers(self, *, limit: int | None = None, offset: int | None = None) -> dict[str, Any]:
        params: dict[str, Any] = {}
        if limit is not None:
            params["limit"] = limit
        if offset is not None:
            params["offset"] = offset
        return await self._http.get("/api/node/peers", params or None)

    async def get_config(self) -> dict[str, Any]:
        return await self._http.get("/api/node/config")

    async def wait_for_sync(
        self,
        *,
        interval: float = 2.0,
        timeout: float = 60.0,
    ) -> dict[str, Any]:
        import asyncio

        deadline = time.monotonic() + timeout
        while True:
            status = await self.get_status()
            if status.get("synced"):
                return status
            if time.monotonic() >= deadline:
                raise TimeoutError(f"Node did not sync within {timeout}s")
            await asyncio.sleep(interval)

File: packages/sdk/src/node.ts
/**
 * Node status and management API.
 */
import type { HttpClient, RequestOptions } from './http.js';
import type { NodeStatus, NodePeersResponse, NodeConfig } from './types.js';

export class NodeApi {
  constructor(private readonly http: HttpClient) {}

  /** Get node status (sync state, peers, version, uptime). */
  async getStatus(opts?: RequestOptions): Promise<NodeStatus> {
    return this.http.get<NodeStatus>('/api/node/status', undefined, opts);
  }

  /** List connected peers. */
  async getPeers(opts?: RequestOptions): Promise<NodePeersResponse> {
    return this.http.get<NodePeersResponse>('/api/node/peers', undefined, opts);
  }

  /** Get node configuration. */
  async getConfig(opts?: RequestOptions): Promise<NodeConfig> {
    return this.http.get<NodeConfig>('/api/node/config', undefined, opts);
  }

  /**
   * Wait until the node reports `synced: true`.
   * Polls every `intervalMs` (default 2 000 ms) up to `timeoutMs` (default 60 000 ms).
   */
  async waitForSync(timeoutMs = 60_000, intervalMs = 2_000, opts?: RequestOptions): Promise<NodeStatus> {
    const deadline = Date.now() + timeoutMs;
    while (Date.now() < deadline) {
      const status = await this.getStatus(opts);
      if (status.synced) return status;
      await sleep(intervalMs);
    }
    throw new Error(`Node did not sync within ${timeoutMs}ms`);
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}
